### ISSUE TITLE
All strings should be iterpolated as templates

### ISSUE THREAD

**Issue by kneufeld** [hashicorp/nomad#6366]
This is somewhat related to #5051 but I'd go a step further and say all strings should "do the right thing" when given a template or even just a `${NOMAD_xxx}` env var.

My use case is that we have many tenants, but our app itself is not multi-tenant so we have a job file for each tenant. I use ansible's templating feature to create the job files but it's still lacking.

For example, it would be super great if you could specify the docker image as a consul kv item.

```
job "tenant1" {
  group "${NOMAD_JOB_NAME}" {
    task "${NOMAD_JOB_NAME}-foo" {
      driver = "docker"
      config {
        image = "{{printf "tenant/%s/docker_image" (env "NOMAD_JOB_NAME") | key}}"

        env {
          WHY_NOT = "{{ why can this only be in a template? }}"
        }
      }
    }
  }
}
```

Then at some future time I could just `consul kv put tenant/tenant1/docker_image newer_version` to do my upgrades.

**Comment by nickethier** [comment]
Currently the template stanza uses an entirely different engine [consul-template](https://github.com/hashicorp/consul-template) to do this. We do plan to incorporate similar flexibility once HCL2 has stabilized and we can refactor the jobspec to use it.

**Comment by tgross** [comment]
With HCL2 support having shipped in Nomad 1.0's beta, I think we can close this one out.
